### ISSUE TITLE
If have multiple network adapters. Multiple clients need to manually set different `spring.cloud.sentinel.transport.client-ip`

### ISSUE THREAD

**Issue by Deycoesr** [alibaba/Sentinel#1361]
### Issue Description

If have multiple network adapters. [`com.alibaba.csp.sentinel.util.HostNameUtil#resolveHost()`](https://github.com/alibaba/Sentinel/blob/master/sentinel-core/src/main/java/com/alibaba/csp/sentinel/util/HostNameUtil.java#L44) May get an incorrect IP.
It will cause the sentinel-dashboard server to fail to connect to the client.

### Describe what happened

 Multiple clients need to manually set different `spring.cloud.sentinel.transport.client-ip`. A bit difficult

### Describe what you expected to happen

Is it possible to get `spring.cloud.sentinel.transport.client-ip` through the sentry-dashboard server.

---

### 问题描述

如果有多个网络适配器. [`com.alibaba.csp.sentinel.util.HostNameUtil#resolveHost()`](https://github.com/alibaba/Sentinel/blob/master/sentinel-core/src/main/java/com/alibaba/csp/sentinel/util/HostNameUtil.java#L44) 可能会获得一个不正确的 IP, 会导致 sentinel-dashboard 服务端不能连接到客户端. 

### 发生了

多个客户端都要手动设置不同的 `spring.cloud.sentinel.transport.client-ip`, 有一点难受.

### 期望发生的事情

是否能够通过 sentinel-dashboard 服务端, 获得 `spring.cloud.sentinel.transport.client-ip`.

**Comment by Deycoesr** [comment]
like this [[Fix alibaba#1361] Get `spring.cloud.sentinel.transport.client-ip` through the sentry dashboard server](https://github.com/Deycoesr/Sentinel/commit/4fe2b591c823f034e9b07ab8b4383da6a8721722)

**Comment by cdfive** [comment]
Maybe it's a new idea by adding `TransportConfig.FETCH_FROM_SERVER` and then dashboard get client's IP from `HttpRequest`, and we may think if it's indeed needed, and how about adding this as a configuration instead of removing `HostNameUtil.getIp()` ?

**Comment by Deycoesr** [comment]
this ? [[Fix alibaba#1361] Add TransportConfig item `csp.sentinel.heartbeat.client.ip.fetch.from.server`](https://github.com/Deycoesr/Sentinel/commit/2d95c38ceb80233720911c084a275d9f42a9fecd)

**Comment by jasonjoo2010** [comment]
@sczyh30 
Coming to this point i don't quite know the original purpose specifying the client ip programmatically instead of fetching the address of remote peer on dashboard automatically. (Surely if an address is brought in heartbeat request it can have a higher priority). Is there anything would make things mess?

**Comment by cdfive** [comment]
I think it's reasonable why the client ip is carried in heartbeat message, since the pair(ip and port) can mark a unique machine of one application.Though ip can be got by dashboard automatically, it seems that port can't. One thing needed to be pay attention to is that the port is used by sentinel transport http server, and is not http port or dubbo port of application. 

How dou you see it and any suggestion? @sczyh30

**Comment by jasonjoo2010** [comment]
> I think it's reasonable why the client ip is carried in heartbeat message, since the pair(ip and port) can mark a unique machine of one application.Though ip can be got by dashboard automatically, it seems that port can't. One thing needed to be pay attention to is that the port is used by sentinel transport http server, and is not http port or dubbo port of application.
> 
> How do you see it and any suggestion? @sczyh30

I agree with you and it's quite straightforward introducing a new property named `port`. And we certainly can make it backward compatible.

**Comment by cdfive** [comment]
Now the heartbeat message did has the `port` property.
1. Fetching the address of remote peer on dashboard automatically.
2. Add a config parameter `csp.sentinel.heartbeat.client.ip.fetch.from.server`, if set, fetch the ip by dashboard, else same as now.

Which is better? @jasonjoo2010 @Deycoesr @sczyh30

**Comment by jasonjoo2010** [comment]
> Now the heartbeat message did has the `port` property.
> 
> 1. Fetching the address of remote peer on dashboard automatically.
> 2. Add a config parameter `csp.sentinel.heartbeat.client.ip.fetch.from.server`, if set, fetch the ip by dashboard, else same as now.
> 
> Which is better? @jasonjoo2010 @Deycoesr @sczyh30

Personally if no other bad effect i prefer fetch address on dashboard side if no address specified manually on client side. Thus, no new configuration item and better performance.

**Comment by cdfive** [comment]
> Personally if no other bad effect i prefer fetch address on dashboard side if no address specified manually on client side.

I'm a bit in favor of this method, but feel worried that it seems the previous way to obtain IP has to be removed. Is it useless to obtain client IP and put it into heartbeat message?

Besides, I found Dubbo also has issue about choosing ip when multiple network adapters, maybe we can learn from it.
Here is a discussion:
[https://github.com/apache/dubbo/issues/1557](https://github.com/apache/dubbo/issues/1557)

**Comment by jasonjoo2010** [comment]
> > Personally if no other bad effect i prefer fetch address on dashboard side if no address specified manually on client side.
> 
> I'm a bit in favor of this method, but feel worried that it seems the previous way to obtain IP has to be removed. Is it useless to obtain client IP and put it into heartbeat message?
> 
> Besides, I found Dubbo also has issue about choosing ip when multiple network adapters, maybe we can learn from it.
> Here is a discussion:
> [apache/dubbo#1557](https://github.com/apache/dubbo/issues/1557)
> 
> One answer in the issue which I think is quite reasonable:
> 
> > The deploy environment for dubbo varies, we highly recommend rely on the first two approaches, specifying the ip you want explicitly can be the best way, avoid guessing as it's unpredictable. In a docker environment, bind the docker ip or host ip can depends, leave the choice to users.

Dubbo has this issue and recommends users specifying it because registry doesn't have the ability to achieve such kind of logic. 

And if something may go wrong changing the default behavior i certainly have no further question on introducing a new configuration.

**Comment by Deycoesr** [comment]
> 1. Fetching the address of remote peer on dashboard automatically.

this needs to consider the `csp.sentinel.heartbeat.client.ip` of  lower version client will not work for higher version dashboard.

---

> 2. Add a config parameter `csp.sentinel.heartbeat.client.ip.fetch.from.server`, if set, fetch the ip by dashboard, else same as now.
 
i think this will be better.
